src/jit/importer.cpp

```diff
@@ -17011,18 +17005,10 @@ void Compiler::impInlineRecordArgInfo(InlineInfo*   pInlineInfo,
 #endif // FEATURE_SIMD
     }
 
-    if (curArgVal->gtFlags & GTF_ORDER_SIDEEFF)
-    {
-        // Right now impInlineSpillLclRefs and impInlineSpillGlobEffects don't take
-        // into account special side effects, so we disallow them during inlining.
-        inlineResult->NoteFatal(InlineObservation::CALLSITE_ARG_HAS_SIDE_EFFECT);
-        return;
-    }
-
-    if (curArgVal->gtFlags & GTF_GLOB_EFFECT)
+    if (curArgVal->gtFlags & GTF_ALL_EFFECT)
     {
         inlCurArgInfo->argHasGlobRef = (curArgVal->gtFlags & GTF_GLOB_REF) != 0;
-        inlCurArgInfo->argHasSideEff = (curArgVal->gtFlags & GTF_SIDE_EFFECT) != 0;
+        inlCurArgInfo->argHasSideEff = (curArgVal->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0;
```

Comment:
`GTF_SIDE_EFFECT` does not take into account the `GTF_ORDER_SIDEEFF` bit that is used to mark volatile accesses (among other things). I also realize now that the original code did not exclude `GTF_GLOB_REF` from the test for side effects, so my earlier suggestion was incorrect. The code should simply be

`inlCurArgInfo->argHasSideEff = (curArgVal->gtFlags & GTF_ALL_EFFECT) != 0`

(which is essentially the same as your original change).